File: ext/examples/differentials.rs
//! This prints all the differentials in the resolution.

use ext::{
    chain_complex::{ChainComplex, FreeChainComplex},
    utils::query_module,
};
use sseq::coordinates::BidegreeGenerator;

fn main() -> anyhow::Result<()> {
    ext::utils::init_logging()?;

    let resolution = query_module(None, false)?;

    for b in resolution.iter_stem() {
        for i in 0..resolution.number_of_gens_in_bidegree(b) {
            let gen = BidegreeGenerator::new(b, i);
            let boundary = resolution.boundary_string(gen);
            println!("d x_{gen:#} = {boundary}");
        }
    }
    Ok(())
}

File: ext_py/examples/differentials.py
#!/usr/bin/env python3
"""Print all the differentials in the resolution.

Python port of ext/examples/differentials.rs.
"""

import _query as query
from ext import sseq


def main():
    resolution = query.query_module()

    for b in resolution.iter_stem():
        for i in range(resolution.number_of_gens_in_bidegree(b)):
            g = sseq.BidegreeGenerator(b, i)
            boundary = resolution.boundary_string(g)
            print(f"d x_{g:#} = {boundary}")


if __name__ == "__main__":
    main()
